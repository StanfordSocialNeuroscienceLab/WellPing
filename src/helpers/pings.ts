import { isToday, addDays } from "date-fns";

import { PingEntity } from "../entities/PingEntity";
import { isTimeThisWeekAsync } from "./configFiles";
import { StreamName } from "./types";

export async function getNumberOfPingsForStreamName(
  streamName: StreamName,
): Promise<number> {
  const countRaw: {
    count: number;
  } = await PingEntity.createQueryBuilder()
    .select("COUNT(id) as count")
    .where("streamName = :streamName", { streamName })
    .getRawOne();
  return countRaw.count;
}

type NumbersOfPingsForAllStreamNames = {
  [stream: string /* actually StreamName */]: number;
};
export async function getNumbersOfPingsForAllStreamNames(): Promise<
  NumbersOfPingsForAllStreamNames
> {
  const pingsGroupByStreamNameRaw: {
    streamName: StreamName;
    count: number;
  }[] = await PingEntity.createQueryBuilder()
    .select("streamName, COUNT(id) as count")
    .groupBy("streamName")
    .getRawMany();

  return pingsGroupByStreamNameRaw.reduce((map, row) => {
    map[row.streamName] = row.count;
    return map;
  }, {} as NumbersOfPingsForAllStreamNames);
}

export async function insertPingAsync({
  notificationTime,
  startTime,
  streamName,
}: {
  notificationTime: Date;
  startTime: Date;
  streamName: StreamName;
}): Promise<PingEntity> {
  const newIndex = (await getNumberOfPingsForStreamName(streamName)) + 1;
  const pingId = `${streamName}${newIndex}`;
  const tzOffset = startTime.getTimezoneOffset();

  const pingEntity = new PingEntity();
  pingEntity.id = pingId;
  pingEntity.notificationTime = notificationTime;
  pingEntity.startTime = startTime;
  pingEntity.streamName = streamName;
  pingEntity.tzOffset = tzOffset;
  await pingEntity.save();

  // Make sure state and database are consistent.
  await pingEntity.reload();

  return pingEntity;
}

export async function addEndTimeToPingAsync(
  pingId: string,
  endTime: Date,
): Promise<PingEntity> {
  const ping = await PingEntity.createQueryBuilder()
    .where("id = :pingId", { pingId })
    .getOne();
  if (ping == null) {
    throw new Error(`pingId ${pingId} not found in getPingsAsync.`);
  }

  ping.endTime = endTime;
  await ping.save();

  // Make sure state and database are consistent.
  ping.reload();

  return ping;
}

export async function getLatestPingAsync(): Promise<PingEntity | null> {
  const latestPing = await PingEntity.createQueryBuilder()
    .orderBy("startTime", "DESC")
    .take(1)
    .getOne();
  if (!latestPing) {
    return null;
  }
  return latestPing;
}

export async function getPingsAsync(): Promise<PingEntity[]> {
  const pings = await PingEntity.createQueryBuilder()
    .orderBy("startTime", "ASC")
    .getMany();
  return pings;
}

export async function getTodayPingsAsync(): Promise<PingEntity[]> {
  const allPings = await getPingsAsync();
  const todayPings: PingEntity[] = [];
  // TODO: USE SQL HERE? (MAKE SURE TIMEZONE PROBLEM)
  for (const ping of allPings) {
    if (isToday(ping.notificationTime)) {
      todayPings.push(ping);
    }
    if (ping.notificationTime > addDays(new Date(), 1)) {
      break;
    }
  }
  return todayPings;
}

export async function getThisWeekPingsAsync(): Promise<PingEntity[]> {
  const allPings = await getPingsAsync();
  const thisWeekPings: PingEntity[] = [];
  // TODO: USE SQL HERE? (MAKE SURE TIMEZONE PROBLEM)
  for (const ping of allPings) {
    if (await isTimeThisWeekAsync(ping.notificationTime)) {
      thisWeekPings.push(ping);
    }
    if (ping.notificationTime > new Date()) {
      // Stop when the notification time is in the future.
      break;
    }
  }
  return thisWeekPings;
}
