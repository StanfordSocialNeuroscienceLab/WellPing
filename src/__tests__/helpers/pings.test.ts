import * as DateMock from "jest-date-mock";
import { Connection } from "typeorm";

import {
  getPingsAsync,
  getTodayPingsAsync,
  insertPingAsync,
  addEndTimeToPingAsync,
  getThisWeekPingsAsync,
} from "../../helpers/pings";
import {
  connectTestDatabaseAsync,
  getTestDatabaseFilename,
} from "../data/database_helper";
import { PINGS, PINGS_DICT } from "../data/pings";

let connection: Connection;
const DB_NAME = "pings";
const DB_FILENAME = getTestDatabaseFilename(DB_NAME);
beforeAll(async () => {
  connection = await connectTestDatabaseAsync(DB_FILENAME);

  // Reset database on start.
  await connection.dropDatabase();
  await connection.synchronize();
});
afterAll(async () => {
  const allPings = await getPingsAsync();
  expect(allPings).toMatchSnapshot();

  await connection.close();
});

test("insert pings and set end date", async () => {
  for (let i = 0; i < PINGS.length; i++) {
    const ping = PINGS[i];

    const addedPing = await insertPingAsync({
      notificationTime: ping.notificationTime,
      startTime: ping.startTime,
      streamName: ping.streamName,
    });
    expect(addedPing).toEqual({
      ...ping,
      endTime: null, // Doesn't have end time yet.
    });

    if (ping.endTime) {
      const updatedEndedPing = await addEndTimeToPingAsync(
        ping.id,
        ping.endTime,
      );
      expect(updatedEndedPing).toEqual(ping);
    }

    expect(await getPingsAsync()).toEqual(PINGS.slice(0, i + 1));
  }
});

test("set end date to non-existent ping", async () => {
  // https://github.com/facebook/jest/issues/1700
  expect(
    (async () => {
      await addEndTimeToPingAsync("_OwO_", new Date());
    })(),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"pingId _OwO_ not found in getPingsAsync."`,
  );

  // The pings data should not be changed.
  expect(await getPingsAsync()).toEqual(PINGS);
});

test("get today's ping", async () => {
  DateMock.advanceTo(+new Date("2010-04-30T08:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-01T08:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-01T10:00:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([PINGS_DICT["cat1"]]);

  DateMock.advanceTo(+new Date("2010-05-01T18:58:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([
    PINGS_DICT["cat1"],
    PINGS_DICT["dog1"],
  ]);

  DateMock.advanceTo(+new Date("2010-05-01T22:46:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([
    PINGS_DICT["cat1"],
    PINGS_DICT["dog1"],
    PINGS_DICT["wolf1"],
  ]);

  DateMock.advanceTo(+new Date("2010-05-02T08:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-02T14:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([PINGS_DICT["cat2"]]);

  DateMock.advanceTo(+new Date("2010-05-03T08:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-03T09:30:00Z"));
  expect(await getTodayPingsAsync()).toEqual([PINGS_DICT["cat3"]]);

  DateMock.advanceTo(+new Date("2010-05-05T08:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-10T23:00:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-10T23:59:59Z"));
  expect(await getTodayPingsAsync()).toEqual([PINGS_DICT["cat4"]]);

  DateMock.advanceTo(+new Date("2010-05-11T00:00:01Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2011-01-01T08:08:08Z"));
  expect(await getTodayPingsAsync()).toEqual([]);

  DateMock.clear();
});

test("get this week's ping", async () => {
  // TODO: weekStartsOn IS NOT CONSIDERED HERE.

  DateMock.advanceTo(+new Date("2010-04-30T08:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-01T08:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-01T10:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([PINGS_DICT["cat1"]]);

  DateMock.advanceTo(+new Date("2010-05-01T18:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([
    PINGS_DICT["cat1"],
    PINGS_DICT["dog1"],
  ]);

  DateMock.advanceTo(+new Date("2010-05-01T22:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([
    PINGS_DICT["cat1"],
    PINGS_DICT["dog1"],
    PINGS_DICT["wolf1"],
  ]);

  DateMock.advanceTo(+new Date("2010-05-02T08:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([
    PINGS_DICT["cat1"],
    PINGS_DICT["dog1"],
    PINGS_DICT["wolf1"],
  ]);

  DateMock.advanceTo(+new Date("2010-05-02T15:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([
    PINGS_DICT["cat1"],
    PINGS_DICT["dog1"],
    PINGS_DICT["wolf1"],
    PINGS_DICT["cat2"],
  ]);

  DateMock.advanceTo(+new Date("2010-05-03T08:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-03T10:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([PINGS_DICT["cat3"]]);

  DateMock.advanceTo(+new Date("2010-05-09T23:59:59Z"));
  expect(await getThisWeekPingsAsync()).toEqual([PINGS_DICT["cat3"]]);

  DateMock.advanceTo(+new Date("2010-05-10T00:00:01Z"));
  expect(await getThisWeekPingsAsync()).toEqual([]);

  DateMock.advanceTo(+new Date("2010-05-11T08:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([PINGS_DICT["cat4"]]);

  DateMock.advanceTo(+new Date("2011-01-01T08:08:08Z"));
  expect(await getThisWeekPingsAsync()).toEqual([]);

  DateMock.clear();
});
