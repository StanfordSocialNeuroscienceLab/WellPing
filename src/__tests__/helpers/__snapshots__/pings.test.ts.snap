// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get this week's ping 1`] = `
Array [
  PingEntity {
    "answers": undefined,
    "endTime": 2010-05-01T08:39:40.000Z,
    "id": "cat1",
    "notificationTime": 2010-05-01T08:25:05.000Z,
    "startTime": 2010-05-01T08:30:01.000Z,
    "streamName": "cat",
    "tzOffset": 0,
  },
  PingEntity {
    "answers": undefined,
    "endTime": null,
    "id": "dog1",
    "notificationTime": 2010-05-01T10:20:00.000Z,
    "startTime": 2010-05-01T10:20:05.000Z,
    "streamName": "dog",
    "tzOffset": 0,
  },
  PingEntity {
    "answers": undefined,
    "endTime": 2010-05-01T20:59:20.000Z,
    "id": "wolf1",
    "notificationTime": 2010-05-01T19:00:05.000Z,
    "startTime": 2010-05-01T20:06:21.000Z,
    "streamName": "wolf",
    "tzOffset": 0,
  },
  PingEntity {
    "answers": undefined,
    "endTime": null,
    "id": "cat2",
    "notificationTime": 2010-05-02T13:08:04.000Z,
    "startTime": 2010-05-02T14:59:01.000Z,
    "streamName": "cat",
    "tzOffset": 0,
  },
  PingEntity {
    "answers": undefined,
    "endTime": 2010-05-03T09:26:02.000Z,
    "id": "cat3",
    "notificationTime": 2010-05-03T09:08:07.000Z,
    "startTime": 2010-05-03T09:26:00.000Z,
    "streamName": "cat",
    "tzOffset": 0,
  },
  PingEntity {
    "answers": undefined,
    "endTime": 2010-05-10T23:59:58.000Z,
    "id": "cat4",
    "notificationTime": 2010-05-10T23:08:07.000Z,
    "startTime": 2010-05-10T23:09:09.000Z,
    "streamName": "cat",
    "tzOffset": 0,
  },
]
`;
